daemon/internal/newrelic/app.go

```diff
@@ -338,3 +340,75 @@ func (app *App) Inactive(threshold time.Duration) bool {
 	}
 	return time.Since(app.LastActivity) > threshold
 }
+
+// filter seen php packages data to avoid sending duplicates
+//
+// the `App` structure contains a map of PHP Packages the reporting
+// application has encountered.
+//
+// the map of packages should persist for the duration of the
+// current connection
+//
+// takes the `PhpPackages.data` byte array as input and unmarshals
+// into an anonymous interface array
+//
+// the JSON format received from the agent is:
+// 		[["package_name","version",{}],...]
+//
+// for each entry, assign the package name and version to the `PhpPackagesKey`
+// struct and use the key to verify data does not exist in the map. If the
+// key does not exist, add it to the map and the array of 'new' packages.
+//
+// convert the array of 'new' packages into a byte array representing
+// the expected data that should match input, minus the duplicates.
+func (app *App) filterPhpPackages(data []byte) []byte {
+	if data == nil {
+		return nil
+	}
+
+	var pkgKey PhpPackagesKey
+	var newPkgs []PhpPackagesKey
+	var x []interface{}
+
+	err := json.Unmarshal(data, &x)
+	if nil != err {
+		log.Errorf("failed to unmarshal php package json: %s", err)
+		return nil
+	}
+
+	for _, pkgJson := range x {
+		pkg, _ := pkgJson.([]interface{})
+		if len(pkg) != 3 {
+			log.Errorf("invalid php package json structure: %+v", pkg)
+			return nil
+		}
+		pkgKey = PhpPackagesKey{pkg[0].(string), pkg[1].(string)}
+		_, ok := app.PhpPackages[pkgKey]
+		if !ok {
+			app.PhpPackages[pkgKey] = struct{}{}
+			newPkgs = append(newPkgs, pkgKey)
+		}
+	}
+
+	if newPkgs == nil {
+		return nil
+	}
+
+	buf := &bytes.Buffer{}
+	buf.WriteString(`[`)
+	for _, pkg := range newPkgs {
+		buf.WriteString(`["`)
+		buf.WriteString(pkg.Name)
+		buf.WriteString(`","`)
+		buf.WriteString(pkg.Version)
+		buf.WriteString(`",{}],`)
+	}
+
+	resJson := buf.Bytes()
+
+	// swap last ',' character with ']'
```

Comment:
What if instead of depending on the format of the JSON from the former block of code we change that block of code to do something like:
```
first = true
for _, pkg := range newPkgs {
  if !first {
     buf.WriteString(`,`)
  }
  <code to create JSON for this pkg>
  first = false
}
```

Just thinking if the code generates the desired format from the start it makes the rest of the code simpler.

Reply:
I prefer not to introduce extra logic unnecessarily. This works, and for the amount of data we expect here, is trivial in terms of speed and space. This is something that should probably be addressed at the agent level to not send JSON at all in another PR.